### PR TITLE
Working with java synchronization

### LINKED ISSUE #5936: Working with java synchronization
#### Proposed title of article
[Languages] Working with java synchronization

#### Proposed article introduction
The synchronization mechanism in java prohibits multiple threads from accessing shared resources. Shared resources include external files, class variables, and database information.
Synchronization is prevalent in multithreaded programming. Synchronized is the keyword that allows your code to run on a single thread at a time without interruption.

#### Key takeaways 
After reading this article, the reader should be able to understand:
-  What is synchronization in java and why do we need it.
- How synchronization works
- How to declare a method as synchronized
- Synchronized method vs Synchronized block 

#### Article quality
This will be an in-depth article covering all the aspects of this concept. It will explain key concepts to readers which is unavailable in the documentations and other platforms. I aim at making it more beginner-friendly simple and easy to understand. More on the coverage will be on the types of synchronization as well as multi-threading with and without synchronization

#### References
N/A

### PR DESCRIPTION
**Pull request by gracy20**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes
This closes #5936

### CODE REVIEW

**Comment by gracy20** [comment]
hello @john-amiscaray, I have implemented the requested changes. please check

File: content/articles/working-with-java-synchronization/index.md [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+### Understanding synchronized methods and synchronized blocks
+To understand the two, let us first look at each separately:
+
+#### Synchronized methods
+These consist of the following properties:
+
+- A synchronized method locks down the whole object. No other thread may access synchronized methods in the object while the method is executing. In non-static methods, this locks them whereas static methods are locked by their class.
+- Synchronized methods need locks on the current object or, in the case of static methods, the whole class. This is because the lock is obtained when the thread enters and released when the thread quits (naturally or by throwing an Exception).
+- Synchronized methods keep the lock for the method scope.
+- A synchronized static function prevents the class object from being changed. A class is a blueprint or template from which objects are built. As a result, an object is a class's instance(result).
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:145]: Here, when I said, *"Not sure what you mean by class object"* I didn't mean that you should explain the concept of classes and objects (the reader should already understand that given the skill level of this tutorial). I just felt that the wording was a little off when you said *class object*. I presume you meant **instances of the class.**

Comment (john-amiscaray) [reply content/articles/working-with-java-synchronization/index.md:145]: I'll change this accordingly. Let me know if I'm wrong.

File: content/articles/working-with-java-synchronization/index.md [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+#### Synchronized blocks
+The following is all that it entails:
+
+- This keyword prevents synchronized blocks from being added to the item between the parentheses. Locked items are only accessible to the threads that are running in the synchronized block.
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md]: You should probably show what the synchronized block looks like. Not showing it here makes this hard to understand

Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md]: It's not clear what you mean by: "prevents synchronized blocks from being added to the item between parentheses"

File: content/articles/working-with-java-synchronization/index.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+- Synchronized blocks utilize it as a lock. When a method is marked as synchronized, the thread owns the monitor or lock object. In this case, you are blocked until the other thread releases the monitor.
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:151]: Still, it isn't clear what you are referring to as **it** in the sentence: "Synchronized blocks utilize it as a lock"

File: content/articles/working-with-java-synchronization/index.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+The output of the thread is : Thread-0 - 3
+The output of the thread is : Thread-1 - 3
+The output of the thread is : Thread-2 - 3
+The output of the thread is : Thread-4 - 5
+The output of the thread is : Thread-3 - 4
+```
+
+As shown above, the threads are chosen at random, and the value is incorrect. The value should rise by 1 but that is not the case. Usually, the output is 3. Threads 0, 1, and 2 share the same value thus showing a race condition. After understanding what a race condition is, let's now look at how to refrain from it.
+
+It is evident that the crucial element (code that changes shared resources) must be limited. Java's synchronized keyword synchronizes shared resource access. This prevents thread interference during atomic operations.
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md]: Please explain what an atomic operation is

**Review by john-amiscaray — CHANGES_REQUESTED** [review]
Please implement these suggestions. I'll keep the pull request open since I don't think these changes are large enough to warrant closing the pull request. Please be diligent however to implement these changes as not to take up space in the queue.

File: content/articles/working-with-java-synchronization/index.md (hunk 1 above)
Comment (gracy20) [reply content/articles/working-with-java-synchronization/index.md]: I think it sounds great.

**Comment by gracy20** [comment]
Hello @john-amiscaray I have implemented the requested changes.

**Comment by john-amiscaray** [comment]
@gracy20 I made some edits just now but I'm not yet done with my current review. You'll know when I'm done once you see my feedback comments.

File: content/articles/working-with-java-synchronization/index.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    public class TestSynchronizedBlock1{    
+    public static void main(String args[]){    
+    Table obj = new Table();//only one object    
+    MyThread1 t1=new MyThread1(obj);    
+    MyThread2 t2=new MyThread2(obj);    
+    t1.start();    
+    t2.start();    
+    }    
+    }    
+```
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md]: There seems to be a race condition issue with this code. Occasionally, I noticed it would output:

```
100
200
300
400
500
5
10
15
20
25
```
Please fix it

Comment (john-amiscaray) [reply content/articles/working-with-java-synchronization/index.md]: Capture

File: content/articles/working-with-java-synchronization/index.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+public class SynchImp {
+	public static void main(String[] args) {
+
+		Implementation b1 = new Implementation();
+		OurThreadExample mt1 = new OurThreadExample(b1, "SECTION");
+		OurThreadExample mt2 = new OurThreadExample(b1, "ENGINEERING");
+		mt1.start();
+		mt2.start();
+	}
+}
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md]: This code also suffers from a race condition error. Occasionally, I will get this output:

Capture

File: content/articles/working-with-java-synchronization/index.md [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+
+public class TestSynchronizedBlock1{
+    public static void main(String[] args){
+        Table obj = new Table();//only one object
+        MyThread1 t1=new MyThread1(obj);
+        MyThread2 t2=new MyThread2(obj);
+        t1.start();
+        t2.start();
+    }
+}
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:204]: This code suffers from a race condition. Occasionally, it will output:

```
100
200
300
400
500
5
10
15
20
25
```

Comment (john-amiscaray) [reply content/articles/working-with-java-synchronization/index.md:204]: This can take 10+ executions to reproduce

File: content/articles/working-with-java-synchronization/index.md [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+public class SynchImp {
+	public static void main(String[] args) {
+
+		Implementation b1 = new Implementation();
+		OurThreadExample mt1 = new OurThreadExample(b1, "SECTION");
+		OurThreadExample mt2 = new OurThreadExample(b1, "ENGINEERING");
+		mt1.start();
+		mt2.start();
+	}
+}
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:280]: This code also suffers from race conditions. Occasionally, it will output:

```
Hello : 
ENGINEERING
Hello : 
ENGINEERING
Hello : 
SECTION
Hello : 
SECTION
```

Comment (john-amiscaray) [reply content/articles/working-with-java-synchronization/index.md:280]: Again, this can take quite a few executions to reproduce.

File: content/articles/working-with-java-synchronization/index.md [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+To provide internal synchronization, Java's lock concept is employed. In Java, each object has its own lock. In this case, the lock idea will be brought into play anytime we utilize the synchronized keyword.
+
+First, a thread with the lock on the object must execute any synchronized methods on it. Post-lock, a thread may invoke any synchronized method on the object. Upon successful completion of the synchronized method, the thread is responsible for releasing the lock. No other threads may perform synchronized methods on the same object while a thread is executing one. However, any non-synchronized procedure may be executed concurrently by the remaining threads. Note that this idea of a lock may be applied on the object level rather than the method level.
+
+Let's look at an example program:
+
+```java
+class Implementation {
+
+	public synchronized void wish(String tag) {
```
Comment (john-amiscaray) [inline content/articles/working-with-java-synchronization/index.md:241]: Inappropriate method name. It seems to be unrelated to the functionality of the method

**Review by john-amiscaray — CHANGES_REQUESTED** [review]
Here are the issues I uncovered from the review that I need you to fix before publication. As a word of advice, you need to be sure to run concurrency-related code many times to look for any race conditions. As you'll see, after running a couple of the code examples many times I uncovered different output than expected. Since we can't predict the JVM's thread scheduling, you might be in situations like this where the output can vary over a smaller percentage of executions. Concurrency-related bugs can therefore be very tricky.

**Comment by john-amiscaray** [comment]
Because I deemed the issues to be large, I will be closing this pull request while you fix them. When you are done, you may create a new pull request and ping me in the comments of it for me to continue the review.

File: content/articles/working-with-java-synchronization/index.md [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+The output of the thread is : Thread-0 - 3
+The output of the thread is : Thread-1 - 3
+The output of the thread is : Thread-2 - 3
+The output of the thread is : Thread-4 - 5
+The output of the thread is : Thread-3 - 4
+```
+
+As shown above, the threads are chosen at random, and the value is incorrect. The value should rise by 1 but that is not the case. Usually, the output is 3. Threads 0, 1, and 2 share the same value thus showing a race condition. After understanding what a race condition is, let's now look at how to refrain from it.
+
+It is evident that the crucial element (code that changes shared resources) must be limited. Additionally, with Java's `synchronized` keyword we can synchronize access to the shared resource. This prevents thread interference during atomic operations. The term "atomic operation" refers to a set of operations that are always performed in unison. All of the atomic actions must be completed at the same time, or none of them can be completed at all.
```
[the remark this answers is not in the document]
Comment (gracy20) [reply content/articles/working-with-java-synchronization/index.md:92]: I meant atomic operations cannot be done when there is interference since they all should complete before other processes start. So the synchronized keyword synchronizes the access to the shared resource.

[the remark this answers is not in the document]
Comment (john-amiscaray) [reply content/articles/working-with-java-synchronization/index.md:92]: I'm not sure what you mean by this. How does the synchronized keyword and how it synchronizes access to the shared resource relate to atomic operations?
